parsing/pprintast.ml

```diff
@@ -1000,6 +1000,11 @@ and module_type ctxt f x =
         pp f "@[<hov2>module@ type@ of@ %a@]" (module_expr ctxt) me
     | Pmty_extension e -> extension ctxt f e
 
+and module_type1 ctxt f x =
+  match x.pmty_desc, x.pmty_attributes with
+  | Pmty_functor _, [] -> paren true (module_type ctxt) f x
+  | _ -> module_type ctxt f x
```

Comment:
Can you explain why you add parentheses around a functor only when there are no attributes, aren't they also necessary if there are some attributes?

Reply:
This is standard in `foo1` functions (see for example [core_type1](https://github.com/314eter/ocaml/blob/9124ce56f95e2480a50062f8897fc80a684f14fc/parsing/pprintast.ml#L273-L274), to defer to the toplevel `foo` function in case of attributes, because it prints the attributes with a lot of parentheses for safety -- see [attribute printing in module_type](https://github.com/314eter/ocaml/blob/9124ce56f95e2480a50062f8897fc80a684f14fc/parsing/pprintast.ml#L957-L959).

On the other hand, the structure of `module_type1` is non-standard: these higher-level printers should handle the constructions that they know do *not* need parethesizing, and defer the others to higher-level printers *or* add parentheses to go back to the toplevel printer (`module_type` here). This is not what this function does, and I am not sure it is correct (for example, it would be wrong if given `(S1 -> S2) with eqn` as argument; but this construction is not type-correct).

I would expect the following two levels:
- `simple_module_type` which prints all constructions that are fully parenthesized (`sig..end`) or atomic (identifiers), for others it calls `module_type` and wraps parentheses around it
- `module_type` which prints the rest, and in particular the `Pmty_functor` construction; in the `S1 -> S2` case, it calls `simple_module_type` on the left-hand-side.
Maybe it would be more correct

Reply:
Thanks, I had missed that.